File: python/test/t_LinearModel_std.py
#! /usr/bin/env python

from __future__ import print_function
from openturns import *
from math import *

TESTPREAMBLE()

try:
    # TEST NUMBER ZERO : DEFAULT & COPY CONSTRUCTORs AND STRING CONVERTER
    print(
        "test number zero : default & copy constructors and string converter")

    vectR = Point(1, 12.0)
    testSample = Sample(1, 1)

    begin = -1.0
    end = 36.92
    testInterval = Interval([begin], [end])
    testInterval.setName("testInterval")

    valColl = ScalarCollection()
    valColl.add(1.0)

    # Constructor from Point
    # Y = Somme (ai * Xi), i=0, n
    # linearModel : (ai), i=0, n
    LM = LinearModel(vectR, testInterval, valColl)
    emptyLM = LinearModel(vectR)

    # String converter #
    print("LM = ", repr(LM))
    print("emptyLM = ", repr(emptyLM))

    # * TEST NUMBER ONE : GET ELEMENTS #
    print("test number one : get elements")

    # coefficients ai
    print("LM.getRegression = ", repr(LM.getRegression()))
    # confidence intervals of the ai coefficients
    print("LM.getConfidenceIntervals = ", repr(LM.getConfidenceIntervals()))
    # p values of the (n+1) coefficients ai:
    print("LM.getPValues = ", repr(LM.getPValues()))

    # * TEST NUMBER TWO : GET PREDICT/RESIDUAL #
    print("test number two : get predicted/residual")
    lmfact = LinearModelFactory()
    size = 20
    oneSample = Sample(size, 1)
    twoSample = Sample(size, 1)
    for i in range(size):
        oneSample[i] = Point(
            1, 7.0 * sin(-3.5 + (6.5 * i) / (size - 1.0)) + 2.0)
        twoSample[i] = Point(
            1, -2.0 * oneSample[i, 0] + 3.0 + 0.05 * sin(oneSample[i, 0]))
    test = lmfact.build(oneSample, twoSample)
    # evaluate the predictions on sample : predict = Somme(aiXi)
    print("LM.getPredicted = ", repr(test.getPredicted(oneSample)))
    # get the residuals epsiloni
    print("LM.getResidual = ", repr(test.getResidual(oneSample, twoSample)))

except:
    import sys
    print("t_LinearModel_std.py", sys.exc_info()[0], sys.exc_info()[1])

File: lib/test/t_LinearModel_std.cxx
//                                               -*- C++ -*-
/**
 *  @brief The test file of class LinearModel
 *
 *  Copyright 2005-2018 Airbus-EDF-IMACS-Phimeca
 *
 *  This library is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public License
 *  along with this library.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
#include "openturns/OT.hxx"
#include "openturns/OTtestcode.hxx"

using namespace OT;
using namespace OT::Test;

int main(int argc, char *argv[])
{
  TESTPREAMBLE;
  OStream fullprint(std::cout);
  setRandomGenerator();

  /** TEST NUMBER ZERO : DEFAULT & COPY CONSTRUCTORs AND STRING CONVERTER */
  fullprint << "test number zero : default & copy constructors and string converter" << std::endl;

  Point vectR(1, 12.0);
  Sample testSample(1, 1);

  Scalar begin = -1.0;
  Scalar end = 36.92;
  Interval testInterval(Point(1, begin), Point(1, end));
  testInterval.setName("testInterval");

  OT::Collection<Scalar> valColl;
  valColl.add(1.0);

  /* Default constructor */
  LinearModel LM(vectR, testInterval, valColl);
  LinearModel LMcopy(LM);

  /* Constructor from Point */
  LinearModel emptyLM(vectR);

  /* String converter */
  fullprint << "LM = " << LM << std::endl;
  fullprint << "LMcopy = " << LMcopy << std::endl;
  fullprint << "emptyLM = " << emptyLM << std::endl;


  /** TEST NUMBER ONE : GET ELEMENTS */
  fullprint << "test number one : get elements" << std::endl;

  fullprint << "LM.getRegression = " << LM.getRegression() << std::endl;
  fullprint << "LM.getConfidenceIntervals = " << LM.getConfidenceIntervals() << std::endl;
  fullprint << "LM.getPValues = " << LM.getPValues() << std::endl;


  /** TEST NUMBER TWO : GET PREDICT/RESIDUAL */
  fullprint << "test number two : get predicted/residual" << std::endl;
  LinearModelFactory lmfact;
  UnsignedInteger size = 20;
  Sample oneSample(size, 1);
  Sample twoSample(size, 1);
  for (UnsignedInteger i = 0; i < size; ++i)
  {
    oneSample[i][0] = 7.0 * sin(-3.5 + (6.5 * i) / (size - 1.0)) + 2.0;
    twoSample[i][0] = -2.0 * oneSample[i][0] + 3.0 + 0.05 * sin(oneSample[i][0]);
  }
  LinearModel test(lmfact.build(oneSample, twoSample));
  fullprint << "LM.getPredicted = " << test.getPredicted(oneSample) << std::endl;
  fullprint << "LM.getResidual = " << test.getResidual(oneSample, twoSample) << std::endl;


  return ExitCode::Success;
}
